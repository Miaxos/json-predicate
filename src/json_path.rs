use json_pointer::{JsonPointer, ParseError};
use serde::{Deserialize, Deserializer, Serialize};
use std::ops::Add;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSONPath(JsonPointer<String, Vec<String>>);

impl Default for JSONPath {
    fn default() -> Self {
        Self(JsonPointer::new(Vec::new()))
    }
}

#[derive(Debug, thiserror::Error)]
pub enum JSONPathError {
    /// An invalid escape sequence was encountered, either a `~` escape or a
    /// `%` escape.
    #[error("An invalid escape sequence was encountered, either a `~` escape or a `%` escape.")]
    InvalidEscape,
    /// An error caused by not having a leading slash on the JSON pointer.
    ///
    /// For example, the string `a/b/c` is not a valid JSON pointer, while
    /// `/a/b/c` is.
    #[error("An error caused by not having a leading slash on the JSON pointer.")]
    NoLeadingSlash,
}

impl JSONPath {
    pub fn new<S: AsRef<str>>(value: S) -> Result<JSONPath, JSONPathError> {
        let ptr = value.as_ref().parse::<JsonPointer<_, _>>();
        match ptr {
            Ok(ptr) => Ok(JSONPath(ptr)),
            Err(ParseError::NoLeadingSlash) => Err(JSONPathError::NoLeadingSlash),
            Err(ParseError::InvalidEscape(_)) => Err(JSONPathError::InvalidEscape),
        }
    }

    pub fn empty() -> Self {
        Self::new("/").expect("Can't fail")
    }

    pub fn take(self) -> JsonPointer<String, Vec<String>> {
        self.0
    }
}

impl Serialize for JSONPath {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let val = self.0.to_string();
        Ok(serializer.serialize_str(&val)?)
    }
}

impl Add for JSONPath {
    type Output = Self;

    fn add(self, other: Self) -> Self::Output {
        // TODO: Fix it because it's an unwrap that shouldn't be here as we should be sure that it
        // works.
        Self::new(format!("{}{}", self.0.to_string(), other.0.to_string())).unwrap()
    }
}

impl<'de> Deserialize<'de> for JSONPath {
    fn deserialize<D>(deserializer: D) -> Result<JSONPath, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        let ptr = s
            .parse::<JsonPointer<_, _>>()
            .map_err(|e| serde::de::Error::custom(format!("{:?}", e)))?;
        Ok(JSONPath(ptr))
    }
}
