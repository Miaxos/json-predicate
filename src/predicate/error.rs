use std::fmt::Display;

use json_pointer::IndexError;

#[derive(Debug)]
pub struct PredicateError {
    kind: PredicateErrorKind,
}

impl PredicateError {
    pub fn unimplemented() -> Self {
        Self {
            kind: PredicateErrorKind::Unimplemented,
        }
    }
    pub fn ty_issue() -> Self {
        Self {
            kind: PredicateErrorKind::IncorrectType,
        }
    }
}

#[derive(Debug)]
enum PredicateErrorKind {
    Unimplemented,
    /// The Predicate Object specifies an unknown predicate operation.
    #[allow(dead_code)]
    UnknownPredicateOperation,
    /// The Predicate Object specifies a JSON Pointer referencing a value that
    /// does not exist and the specified Predicate operation is not
    /// specifically intended to test for the absence of a value
    /// (i.e. the "undefined" and "defined" predicates)
    JsonPointerIssue {
        issue: IndexError,
    },
    /// A First Order Predicate Object specifies a predicate operation that
    /// requires a "value" member providing the condition to test but no
    /// "value" member is provided.
    #[allow(dead_code)]
    NoValueMember,
    /// The "value" member given for a given predicate operation is of an
    /// unexpected or unsupported type for that operation
    /// (e.g. specifying a string value for the "more" and "less" predicate operations).
    IncorrectType,
}

impl Display for PredicateErrorKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnknownPredicateOperation { .. } => f.write_str("UnknownPredicateOperation"),
            Self::JsonPointerIssue { issue } => match issue {
                IndexError::NoSuchKey(key) => f.write_fmt(format_args!(
                    "The pointer pointed to a nonexistent key, pointed key: {}",
                    key
                )),
                IndexError::NotIndexable => {
                    f.write_str("The pointer resulted in trying to index a non-indexable type")
                }
                IndexError::OutOfBounds(idx) => f.write_fmt(format_args!(
                    "The pointer pointed to an out-of-bounds value in an array, pointed at {}",
                    idx
                )),
            },
            Self::NoValueMember { .. } => f.write_str("NoValueMember"),
            Self::IncorrectType { .. } => f.write_str("IncorrectType"),
            Self::Unimplemented { .. } => f.write_str("Unimplemented"),
        }
    }
}

impl std::error::Error for PredicateError {}

impl Display for PredicateError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{}", self.kind))
    }
}

impl From<IndexError> for PredicateError {
    fn from(value: IndexError) -> Self {
        PredicateError {
            kind: PredicateErrorKind::JsonPointerIssue { issue: value },
        }
    }
}
