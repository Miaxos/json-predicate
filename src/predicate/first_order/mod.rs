use crate::Predicate;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use self::{
    contains::Contains, defined::Defined, end::End, less::Less, matches::Matches, more::More,
    r#in::In, r#type::Type, start::Start, test::Test, undefined::Undefined,
};

use super::{context::PredicateContext, error::PredicateError, PredicateImpl};

pub mod contains;
pub mod defined;
pub mod end;
pub mod r#in;
pub mod less;
pub mod matches;
pub mod more;
pub mod start;
pub mod test;
pub mod r#type;
pub mod undefined;

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
#[serde(untagged)]
pub enum FirstOrder {
    Contains(Contains),
    Defined(Defined),
    Undefined(Undefined),
    Start(Start),
    End(End),
    Type(Type),
    In(In),
    Test(Test),
    Matches(Matches),
    Less(Less),
    More(More),
    // Contained,
    // Intersects,
}

impl PredicateImpl for FirstOrder {
    fn evaluate(&self, data: &Value, ctx: PredicateContext) -> Result<bool, PredicateError> {
        match self {
            FirstOrder::Contains(contains) => contains.evaluate(data, ctx),
            FirstOrder::Defined(defined) => defined.evaluate(data, ctx),
            FirstOrder::Undefined(undefined) => undefined.evaluate(data, ctx),
            FirstOrder::Start(start) => start.evaluate(data, ctx),
            FirstOrder::End(end) => end.evaluate(data, ctx),
            FirstOrder::Type(ty) => ty.evaluate(data, ctx),
            FirstOrder::In(ty) => ty.evaluate(data, ctx),
            FirstOrder::Test(ty) => ty.evaluate(data, ctx),
            FirstOrder::Matches(ty) => ty.evaluate(data, ctx),
            FirstOrder::Less(ty) => ty.evaluate(data, ctx),
            FirstOrder::More(ty) => ty.evaluate(data, ctx),
        }
    }
}

impl From<FirstOrder> for Predicate {
    fn from(value: FirstOrder) -> Self {
        Predicate::FirstOrder(value)
    }
}

#[cfg(test)]
mod tests {
    use serde::Deserialize;

    use crate::predicate::first_order::FirstOrder;

    #[test]
    fn deser_test() {
        let contains_expect = serde_json::json!({
             "op": "contains",
             "path": "/a/b",
             "value": " is a "
        });

        let deser = FirstOrder::deserialize(contains_expect).unwrap();

        insta::assert_debug_snapshot!(deser, @r###"
        Contains(
            Contains {
                path: Some(
                    JSONPath(
                        JsonPointer {
                            ref_toks: [
                                "a",
                                "b",
                            ],
                            _phantom: PhantomData<alloc::string::String>,
                        },
                    ),
                ),
                ignore_case: false,
                value: String(
                    " is a ",
                ),
            },
        )
        "###);
    }
}
