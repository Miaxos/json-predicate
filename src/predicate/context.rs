use std::error::Error;

use crate::json_path::{JSONPath, JSONPathError};

/// The context to run the evaluation, you can have an empty context with
/// `PredicateContext::default()`.
#[derive(Default, Clone, Debug)]
pub struct PredicateContext {
    location: Option<JSONPath>,
}

#[derive(Debug, thiserror::Error)]
pub enum PredicateContextError {
    #[error("{0}")]
    Deserialize(#[from] Box<dyn Error>),
    #[error("{0}")]
    JSONPath(#[from] JSONPathError),
}

impl From<JSONPath> for PredicateContext {
    fn from(value: JSONPath) -> Self {
        Self {
            location: Some(value),
        }
    }
}

impl From<Option<JSONPath>> for PredicateContext {
    fn from(value: Option<JSONPath>) -> Self {
        Self { location: value }
    }
}

impl PredicateContext {
    pub fn new(path: String) -> Result<Self, PredicateContextError> {
        let location = JSONPath::new(path)?;

        Ok(Self {
            location: Some(location),
        })
    }

    pub fn final_path(&self, path: &Option<JSONPath>) -> Option<JSONPath> {
        // TODO: Remove cloning
        match (self.location.clone(), path.clone()) {
            (None, None) => None,
            (Some(path), None) | (None, Some(path)) => Some(path),
            (Some(root), Some(path)) => Some(root + path),
        }
    }
}
