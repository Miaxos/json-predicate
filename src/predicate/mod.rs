use serde::{Deserialize, Serialize};

pub mod first_order;
use first_order::FirstOrder;
use serde_json::Value;

use self::{context::PredicateContext, error::PredicateError, second_order::SecondOrder};

pub mod context;
pub mod second_order;

mod error;

/// A Predicate is an Object whose members describe a testable
/// condition that evaluates as either true or false.
///
/// The essential components of a Predicate include:
///
/// - A label identifying the predicate operation,
/// - A reference to the value being tested, and
/// - The condition against which the referenced value is to be evaluated.
///
/// Predicate objects MUST have exactly one "op" member whose value
/// indicates the type of predicate operation to perform.  It's value
/// MUST be one of: "and", "contains", "contains-", "defined", "ends",
/// "ends-", "in", "in-", "less", "matches", "matches-", "more", "not",
/// "or", "starts", "starts-", "test", "test-", "type", or "undefined".
///
/// The semantics for each are defined in the sections that follow.
///
/// Note that the value of the "op" member is case-sensitive and that
/// each of the operations listed are in lower-case.  The value "Starts",
/// for example, is not equivalent to "starts".
///
/// If the "op" member specifies any value other than one of those listed
/// above, the evaluation of the predicate operation MUST cease and be
/// handled as if a boolean value of "false" was returned.  The
/// application processing the predicate operation MAY signal that an
/// error condition has occurred depending on the specific requirements
/// of the application within which JSON Predicates are being used.
///
/// The remaining structure of each predicate operation depends on the
/// specific type.  There are two basic types of predicates.
///
/// -  First Order Predicates that are used to test one name value pair
///    against a single condition, and
///
/// -  Second Order Predicates that aggregate one or more subordinate
///    First or Second Order Predicates.
///
/// In addition to the required "op" member, First Order Predicates have
/// exactly one "path" member whose value MUST be a string containing a
/// JSON-Pointer [RFC6901] value referencing the name value pair that is
/// to be tested.  If the "path" member is not specified within the
/// predicate object, it's value is assumed to be an empty string.
///
/// Second Order Predicates MUST have exactly one "apply" member whose
/// value is a JSON Array containing one or more First or Second Order
/// Predicate Objects.
///
///
/// Additional members can be required depending on the specific
/// predicate operation.  All other members not explicitly defined by
/// this specification MUST be ignored.
///
/// Note that the ordering of members in JSON objects is not significant;
/// therefore the following operations are equivalent:
///
/// ```json
/// {"op": "contains", "path": "/a/b/c", "value": "ABC"}
/// {"path": "/a/b/c", "op": "contains", "value": "ABC"}
/// {"value": "ABC", "path": "/a/b/c", "op": "contains"}
/// ```
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
#[serde(untagged)]
pub enum Predicate {
    FirstOrder(FirstOrder),
    SecondOrder(SecondOrder),
}

impl PredicateImpl for Predicate {
    fn evaluate(&self, data: &Value, ctx: PredicateContext) -> Result<bool, PredicateError> {
        match self {
            Predicate::FirstOrder(fo) => fo.evaluate(data, ctx),
            Predicate::SecondOrder(fo) => fo.evaluate(data, ctx),
        }
    }
}

pub trait PredicateImpl {
    /// Evaluate the predicate against the provided JSON
    ///
    /// The error result means the predicate couldn't be evaluated properly so
    /// you have to assume it's a false result and you'll also have more information about why it
    /// didn't validate.
    fn evaluate(&self, data: &Value, ctx: PredicateContext) -> Result<bool, PredicateError>;
}
