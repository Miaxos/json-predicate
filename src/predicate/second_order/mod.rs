use crate::Predicate;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use self::{and::And, not::Not, or::Or};

use super::{context::PredicateContext, error::PredicateError, PredicateImpl};

pub mod and;
pub mod not;
pub mod or;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
#[serde(untagged)]
pub enum SecondOrder {
    And(And),
    Or(Or),
    Not(Not),
}

impl PredicateImpl for SecondOrder {
    fn evaluate(&self, data: &Value, ctx: PredicateContext) -> Result<bool, PredicateError> {
        match self {
            Self::And(ty) => ty.evaluate(data, ctx),
            Self::Or(ty) => ty.evaluate(data, ctx),
            Self::Not(ty) => ty.evaluate(data, ctx),
        }
    }
}

impl From<SecondOrder> for Predicate {
    fn from(value: SecondOrder) -> Self {
        Predicate::SecondOrder(value)
    }
}

#[cfg(test)]
mod tests {
    use serde::Deserialize;

    use crate::predicate::first_order::FirstOrder;

    #[test]
    fn deser_test() {
        let contains_expect = serde_json::json!({
             "op": "contains",
             "path": "/a/b",
             "value": " is a "
        });

        let deser = FirstOrder::deserialize(contains_expect).unwrap();

        insta::assert_debug_snapshot!(deser, @r###"
        Contains(
            Contains {
                path: Some(
                    JSONPath(
                        JsonPointer {
                            ref_toks: [
                                "a",
                                "b",
                            ],
                            _phantom: PhantomData<alloc::string::String>,
                        },
                    ),
                ),
                ignore_case: false,
                value: String(
                    " is a ",
                ),
            },
        )
        "###);
    }
}
