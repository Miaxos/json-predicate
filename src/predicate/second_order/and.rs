use derive_builder::Builder;
use serde_json::Value;
use std::marker::PhantomData;

use serde::de::{self, Deserialize, Deserializer, MapAccess, Visitor};
use serde::{ser::SerializeStruct, Serialize};

use crate::json_path::JSONPath;
use crate::predicate::context::PredicateContext;
use crate::predicate::error::PredicateError;
use crate::predicate::PredicateImpl;
use crate::Predicate;

use super::SecondOrder;

#[derive(Debug, Clone, PartialEq, Eq, Builder)]
#[builder(pattern = "owned", setter(into, strip_option))]
pub struct And {
    /// Must be a JSON Pointer
    /// https://tools.ietf.org/html/rfc6901
    /// If the "path" member is not specified within the predicate object, it's value is assumed to be an empty string.
    #[builder(default)]
    pub path: Option<JSONPath>,
    pub apply: Vec<Predicate>,
}

impl Serialize for And {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("And", 3)?;
        state.serialize_field("op", "and")?;
        state.serialize_field("path", &self.path)?;
        state.serialize_field("apply", &self.apply)?;
        state.end()
    }
}

impl<'de> Deserialize<'de> for And {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[allow(non_camel_case_types)]
        enum Field {
            op,
            path,
            apply,
            __ignore,
        }
        struct FieldVisitor;

        impl<'de> Visitor<'de> for FieldVisitor {
            type Value = Field;
            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("field identifier")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                match value {
                    "op" => Ok(Field::op),
                    "apply" => Ok(Field::apply),
                    "path" => Ok(Field::path),
                    _ => Ok(Field::__ignore),
                }
            }
        }

        impl<'de> Deserialize<'de> for Field {
            #[inline]
            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
            where
                D: Deserializer<'de>,
            {
                Deserializer::deserialize_identifier(deserializer, FieldVisitor)
            }
        }

        struct VisitorAnd<'de> {
            marker: PhantomData<And>,
            lifetime: PhantomData<&'de ()>,
        }

        impl<'de> Visitor<'de> for VisitorAnd<'de> {
            type Value = And;
            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("And")
            }

            #[inline]
            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
            where
                A: MapAccess<'de>,
            {
                let mut path: Option<Option<JSONPath>> = None;
                let mut apply: Option<Vec<Predicate>> = None;
                let mut op: Option<String> = None;

                while let Some(key) = MapAccess::next_key::<Field>(&mut map)? {
                    match key {
                        Field::op => {
                            if op.is_some() {
                                return Err(serde::de::Error::duplicate_field("op"));
                            }
                            op = Some(MapAccess::next_value::<String>(&mut map)?);
                        }
                        Field::path => {
                            if path.is_some() {
                                return Err(serde::de::Error::duplicate_field("path"));
                            }
                            path = Some(MapAccess::next_value::<Option<JSONPath>>(&mut map)?);
                        }
                        Field::apply => {
                            if apply.is_some() {
                                return Err(serde::de::Error::duplicate_field("value"));
                            }
                            apply = Some(MapAccess::next_value::<Vec<Predicate>>(&mut map)?);
                        }
                        Field::__ignore => {}
                    }
                }

                let path = path.flatten();
                let apply = apply.ok_or(serde::de::Error::missing_field("apply"))?;
                let op = op.ok_or(serde::de::Error::missing_field("op"))?;

                if op.as_str() != "and" {
                    return Err(serde::de::Error::custom("`op` should be `and`"));
                }

                Ok(And { apply, path })
            }
        }

        const FIELDS: &'static [&'static str] = &["op", "apply", "path"];
        Deserializer::deserialize_struct(
            deserializer,
            "And",
            FIELDS,
            VisitorAnd {
                marker: PhantomData::<And>,
                lifetime: PhantomData,
            },
        )
    }
}

impl From<And> for SecondOrder {
    fn from(value: And) -> Self {
        SecondOrder::And(value)
    }
}

impl PredicateImpl for And {
    fn evaluate(&self, data: &Value, ctx: PredicateContext) -> Result<bool, PredicateError> {
        let path = ctx.final_path(&self.path);
        let ctx = PredicateContext::from(path);

        for ap in &self.apply {
            let pred = ap.evaluate(data, ctx.clone());

            if pred.is_ok() {
                continue;
            } else {
                return pred;
            }
        }

        Ok(true)
    }
}

#[cfg(test)]
mod tests {
    use serde::Deserialize;

    use crate::predicate::second_order::and::And;

    #[test]
    fn deser_test() {
        let and_expect = serde_json::json!({
             "op": "and",
             "apply": [
               {
                 "op": "defined",
                 "path": "/a/b"
               },
               {
                 "op": "less",
                 "path": "/a/c/d",
                 "value": 15
               }
            ]
        });

        let deser = And::deserialize(and_expect).unwrap();

        insta::assert_json_snapshot!(deser);
    }
}
