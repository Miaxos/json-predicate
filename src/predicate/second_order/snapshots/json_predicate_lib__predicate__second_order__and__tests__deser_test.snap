---
source: src/predicate/second_order/and.rs
expression: deser

---
{
  "op": "and",
  "path": null,
  "apply": [
    {
      "op": "defined",
      "path": "/a/b"
    },
    {
      "op": "less",
      "path": "/a/c/d",
      "value": 15
    }
  ]
}
